### DISCUSSION TITLE
How to expose Rust trait methods in Python

### DISCUSSION THREAD

**Discussion by azazdeaz** [PyO3/pyo3#3503] [Questions] [answered]
When a struct implements a custom trait, how can i use the functions defined by the trait in Python?

Here is a minimal example:
```rs
trait Bar {
    fn bar(&self) -> i32;
    fn barbar(&self) -> i32 {
        self.bar() * 2
    }
}

#[pyclass]
pub struct Foo {
    pub x: i32,
}

#[pymethods]
impl Foo {
    #[new]
    pub fn new(x: i32) -> Self {
        Self { x }
    }
}


impl Bar for Foo {
    fn bar(&self) -> i32 {
        self.x
    }
}
```
Is it possible to use the `bar()` and `barbar()` methods from python?

**Comment by mejrs** [comment, marked as answer]
No, you have to define methods that forward to the trait. See https://pyo3.rs/v0.19.2/trait_bounds
